MTKA-1302: Add Rich Text repeatable option

nickcernis opened the pull request:
## Description

Rich Text fields now include an optional “Make this field repeatable” checkbox.

It behaves the same way as the Text field's repeatable option, allowing publishers to input multiple rich text fields as separate rows.

https://wpengine.atlassian.net/browse/mtka-1302

## Testing

Includes e2e tests, unit tests for GraphQL/REST, and Jest tests for components.

To test manually:

You may need to first `npm ci` before you `npm start` or `npm run build`. I added `uuid` as a dependency (it was already a sub-dependency of Jest; this just allows us to use safely in front-end code).

1. Create a model with the rich text field.
2. Tick the “make repeatable” option.
3. Add an entry, then enter and save repeatable rich text content.

## Screenshots

### Developer side

<img width="703" alt="rich-text-option-developer-app" src="https://user-images.githubusercontent.com/647669/158813490-ef00dc06-540b-453c-bdda-d4705ceabf53.png">

### Publisher side

<img width="1607" alt="rich-text-repeater-single-field" src="https://user-images.githubusercontent.com/647669/158813516-389dfcf3-d072-42a5-bbb9-9d1d0fea2bff.png">

![rich-text-repeater-multiple-fields](https://user-images.githubusercontent.com/647669/158813456-02d220cf-5e96-4600-960a-fe0e51fed1cf.png)

## Documentation Changes

n/a 

## Dependant PRs

n/a

nickcernis committed:
feat: add Rich Text repeatable option to settings

nickcernis committed:
wip: load rich text editor for repeatables

nickcernis committed:
chore: sanitize rich text arrays

nickcernis committed:
chore: register repeating rich text for GraphQL

nickcernis committed:
feat: add “Add Item” button to rich text repeater

nickcernis committed:
fix: remove broken variable label text

nickcernis committed:
refactor: extract components

nickcernis committed:
deps: save uuid JS dependency

To generate unique IDs for repeatable rich text fields.

We can't use the index of repeatable rich text
fields as their ID, because in a repeatable context
this is not a unique value (an item with index of '3' has
its ID changed to '2' if you delete the item at index '2'
by clicking the “delete item” button).

These IDs need to stay constant and stable regardless
of their position in the list. So we give each a unique ID
for the purpose of initializing TinyMCE wherever they
appear.

The unique ID also doubles as the key for use in React.
This prevents unneeded re-renders when deleting items,
and stops quirks such as deleting an item from the middle
of the list causing the one at the end of the list to be removed.

nickcernis committed:
fix: give Rich Text fields unique IDs

Instead of depending on the index as the key/id.

Prevents issues where the editor would not initialize for newly
added fields.

Also stops React from dropping the last field in the list when
a field in the middle is deleted.

nickcernis committed:
refactor: remove unused prop

Now using custom IDs instead of fieldID-derived values
for repeating rich text fields.

nickcernis committed:
fix: ensure solo rich text fields still load

nickcernis committed:
design: style components with Emotion

nickcernis committed:
refactor: use consistent ID format

So that solo rich text fields use the same uuids as repeatable rich text
fields.

nickcernis committed:
refactor: extract shared Emotion colors

andrewbotz commented on includes/shared-assets/js/emotion.js:
```diff
@@ -0,0 +1,11 @@
+ * Shared styles for components styled with Emotion.
+ */
+
+export const colors = {
```
Love this change and the use of colors object throughout. I was wondering if we could use emotion's ThemeProvider that way we can get all these values plus anything else without having to ever worry about importing colors.

Here's an example : https://github.com/wpengine/atlas-content-modeler/pull/456/files 

Ultimately we could do this for colors, but also common values in relation to other things, such as typography and the alike.

nickcernis replied:
I really like this idea! I hadn't considered using ThemeProvider because we have no plans to offer styling variations just yet, but the API looks clean and it could help us in the future if we need to implement a dark mode etc.

Not sure if it belongs in this PR or is better done as a separate story, though. I went with importing a bare object here just because it's pretty minimal. Open to ideas on that, though. Perhaps I could pull out my color imports and just use bare strings, and we can refactor later?

andrewbotz replied:
Oh ya, totally fine not diving into ThemeProvider now.

andrewbotz commented on includes/publisher/js/src/components/RichText/SoloRichTextEditorField.jsx:
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import RichTextEditorHeader from "./RichTextEditorHeader";
+
+const SoloRichTextEditorField = ({ modelSlug, field, fieldId }) => {
```
Clean! Love it.

nickcernis committed:
fix: remove unused required property

Rich Text area fields currently don't have a required option.

andrewbotz commented on includes/publisher/js/src/components/RichText/RichTextEditor.jsx:
```diff
@@ -0,0 +1,76 @@
+		const initializeEditorWhenReady = () => {
+			/**
+			 * WP defines getDefaultSettings() in an admin footer script tag after
+			 * admin scripts are enqueued, so we must wait for it to be available.
```
Love the comments. Curious, do we need to do this often? As in, set a timeout for WP to finish something?

nickcernis replied:
It's pretty rare! (I think this might be the only place we do it.)

We can usually set a WP core script as a dependency of our script, which ensures it loads before our script has loaded. With scripts that WP core has injected inline, that becomes harder, which is why I went with the workaround here. More details into what it solves and how here: https://github.com/wpengine/atlas-content-modeler/pull/400

andrewbotz replied:
Thank you for the reply and explanation.

nickcernis committed:
refactor: extract editor init as custom hook

To improve readability and add more docs.

nickcernis committed:
refactor: improve hook name

nickcernis committed:
refactor: pass editor ids directly to useWpEditor

nickcernis committed:
refactor: improve var names

nickcernis committed:
refactor: improve add button markup, organisation

nickcernis committed:
fix: remove unneeded space

nickcernis committed:
refactor: improve delete button markup, legibility

nickcernis committed:
refactor: use same global reference style

nickcernis committed:
refactor: remove unneeded preventDefault

Now that type="button" there's no default event to prevent.

nickcernis committed:
fix: use straight quotes

nickcernis committed:
fix: prefix ids

HTML IDs have to start with a letter, not a number.

rfmeier commented:
Your buttons are nice and clean. Would it be worth moving them to a more generic directory? They can easily be reused with the other repeatable fields.

nickcernis committed:
test: e2e test of repeatable rich text fields

nickcernis committed:
refactor: simplify component names

nickcernis committed:
refactor: use more generic testid

To ease button re-use across components and test types.

nickcernis commented:
@rfmeier Sure, I can do that. I think they'll need some light refactoring — we should probably pass an `addItem` and `removeItem` function prop to each (instead of `setValues`) if we want to use them more generically.

That way the button would receive a function that it passes straight to `onClick` instead of holding logic to add/remove rows itself. (I had to adapt the Rich Text field to use uuids for each field instead of a 0-based index, which might not be needed for other fields, so the buttons in this PR currently include uuid logic that others might not need.)

Will take a look tomorrow…

nickcernis committed:
test: update snapshot

nickcernis committed:
refactor: pass add/del function props to buttons

To make them more reusable between components.

nickcernis committed:
refactor: move buttons to shared folder

nickcernis commented:
@rfmeier I refactored the buttons in https://github.com/wpengine/atlas-content-modeler/pull/462/commits/c6d5219b62aed907cc1f20087233e72a3ce05304 and moved them in https://github.com/wpengine/atlas-content-modeler/pull/462/commits/0abdff55d2783ea90a77deaf43bf1a95941a06bc.

Nice idea to make them more reuseable. If you see other ways to improve them feel free to say.

nickcernis committed:
fix: use uuid for single field, pass uuid as prop

nickcernis committed:
test: GraphQL response for rich text repeaters

nickcernis committed:
refactor: use existing acm-icons conventions

nickcernis committed:
docs: improve comment

nickcernis committed:
fix: add React import

Required for the React Test Renderer in Jest tests.

nickcernis committed:
refactor: extract UUID function

To allow it to be mocked more easily with Jest.

We don't want Jest tests to result in true UUIDs because
snapshots comparisons will fail between runs.

nickcernis committed:
test: add Jest tests for Rich Text components

nickcernis committed:
docs: tweak comment

nickcernis committed:
test: basic verification of uuid function

nickcernis committed:
test: check rich text repeaters in rest response

nickcernis committed:
chore: add to changelog

nickcernis committed:
design: reduce default height of TinyMCE instances

At Mason's request in design review.

nickcernis committed:
design: bump height a little

Mason wants three lines to be visible.

rfmeier reviewed (approved):
Excellent. I didn't see any issues. Great work 👍
